File: argon-plugin/src/main/scala/dev/argon/plugin/tube/SerializedTubePlus.scala
package dev.argon.plugin.tube

import dev.argon.tube.*
import zio.*

trait SerializedTubePlus[R, E, ExternMethodImplementation, ExternFunctionImplementation, ExternClassConstructorImplementation] extends SerializedTube[R, E] {
  def getExternMethodImplementation(id: BigInt): ZIO[R, E, ExternMethodImplementation]
  def getExternFunctionImplementation(id: BigInt): ZIO[R, E, ExternFunctionImplementation]
  def getExternClassConstructorImplementation(id: BigInt): ZIO[R, E, ExternClassConstructorImplementation]
  def getVTableDiff(id: BigInt): ZIO[R, E, VTable] 
}

File: plugin/api/java/src/main/java/dev/argon/plugin/api/tube/SerializedTubePlus.java
package dev.argon.plugin.api.tube;

import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.math.BigInteger;

public interface SerializedTubePlus<E extends Throwable, ExternMethodImplementation, ExternFunctionImplementation, ExternClassConstructorImplementation> extends SerializedTube<E> {
    @NotNull ExternMethodImplementation getExternMethodImplementation(@NotNull BigInteger id) throws E, IOException, InterruptedException;
    @NotNull ExternFunctionImplementation getExternFunctionImplementation(@NotNull BigInteger id) throws E, IOException, InterruptedException;
    @NotNull ExternClassConstructorImplementation getExternClassConstructorImplementation(@NotNull BigInteger id) throws E, IOException, InterruptedException;

    @NotNull VTable getVTableDiff(@NotNull BigInteger id) throws E, IOException, InterruptedException;
}
